js.go

```diff
@@ -300,6 +347,304 @@ func (js *js) Publish(subj string, data []byte, opts ...PubOpt) (*PubAck, error)
 	return js.PublishMsg(&Msg{Subject: subj, Data: data}, opts...)
 }
 
+// PubAckFuture is a future for a PubAck.
+type PubAckFuture struct {
+	js     *js
+	msg    *Msg
+	pa     *PubAck
+	st     time.Time
+	err    error
+	errCh  chan error
+	doneCh chan *PubAck
+}
+
+func (paf *PubAckFuture) Ok() <-chan *PubAck {
+	paf.js.mu.Lock()
+	defer paf.js.mu.Unlock()
+
+	if paf.doneCh == nil {
+		paf.doneCh = make(chan *PubAck, 1)
+	}
+	if paf.pa != nil {
+		paf.doneCh <- paf.pa
+	}
+	return paf.doneCh
+}
+
+func (paf *PubAckFuture) Err() <-chan error {
+	paf.js.mu.Lock()
+	defer paf.js.mu.Unlock()
+
+	if paf.errCh == nil {
+		paf.errCh = make(chan error, 1)
+	}
+	if paf.err != nil {
```

Comment:
Think here we need to check if the channel is empty before sending, otherwise we will block when sending until previous consumer is done.
```suggestion
	if paf.err != nil && len(paf.errCh) == 0 {
```

Reply:
We create buffered with size of 1 and will at most only ever send 1 error to this channel.

Reply:
yes but thinking of the case when `Err()` is called twice without but consuming, for example if doing this right now the second call will block. 

```
		first := paf.Err()
		second := paf.Err()
```

With the suggested change then both would get the same buffered channel and the first to consume from it would receive the error, the other one would miss it.

Reply:
ok I understand what you mean now.. I think we should only place the error on there if we create the channel..

Reply:
ok I think I got it now, see CL below.